DESIGN.md

```diff
@@ -0,0 +1,79 @@
+
+To make the abstraction as generic as possible, all of above core traits only define methods to access the address space, and they never define methods to manage (create, delete, insert, remove etc) address spaces. By this way, the address space consumers (virtio device drivers, vhost-user drivers and boot loaders etc) may be decoupled from the address space provider (typically a hypervisor).
+
+### Specialization for Virtual Machine Physical Address Space
+The generic address space crates are specialized to access VM's physical memory with following traits:
+- GuestAddress: represents a guest physical address (GPA). On ARM64, a 32-bit hypervisor may be used to support a 64-bit VM. For simplicity, u64 is used to store the the raw value no matter it is a 32-bit or 64-bit virtual machine.
+- GuestMemoryRegion: used to represent a continuous region of VM's physical memory.
+- GuestMemory: used to represent a collection of GuestMemoryRegion objects. The main responsibilities of the GuestMemory trait are:
+	- hide the detail of accessing VM's physical address (for example complex hierarchical structures).
+	- map a request address to a GuestMemoryRegion object and relay the request to it.
+	- handle cases where an access request spanning two or more GuestMemoryRegion objects.
+
+The VM memory consumers, such as virtio device drivers, vhost drivers and boot loaders etc, should only rely on traits and structs defined here to access VM's physical memory.
+
+### A Sample and Default Backend Implementation Based on mmap()
+Provide a default and sample implementation of the GuestMemory trait by mmapping VM's physical memory into current process. Three data structures are defined here:
+- MmapRegion: mmap a continous range of VM's physical memory into current and provide methods to access the mmapped memory.
+- GuestRegionMmap: a wrapper structure to map VM's physical address into (mmap\_region, offset) tuple.
+- GuestMemoryMmap: manage a collection of GuestRegionMmap objects for a VM.
+
+One of the main responsibilities of the GuestMemoryMmap object is to handle the use cases where an access request crosses the memory region boundary. This scenario may be triggered when memory hotplug is supported. So there's a tradeoff between functionality and code complexity:
+- use following pattern for simplicity which fails when the request crosses region boundary. It's current default behavior in the crosvm and firecracker project.
+```rust
+	let guest_memory_mmap: GuestMemoryMmap = ...
+	let addr: GuestAddress = ...
+        let buf = &mut [0u8; 5];
+	let result = guest_memory_mmap.find_region(addr).unwrap().write_slice(buf, addr);
+```
+- use following pattern for functionality to support request crossing region boundary:
+```rust
+	let guest_memory_mmap: GuestMemoryMmap = ...
+	let addr: GuestAddress = ...
+        let buf = &mut [0u8; 5];
+	let result = guest_memory_mmap.write_slice(buf, addr);
```

Comment:
I think those 2 examples are a little confusing:

### Why is this example crossing the memory region boundary?
We need to be more explicit on how we build `guest_memory_mmap`, maybe something like that (Not built...):

```Rust
const GUEST_START: u64 = 0x1000;
const GUEST_SIZE: usize = 0x2000;

let guest_start_addr = GuestAddress(GUEST_START);
let guest_memory: GuestMemoryMmap::new(&vec![(guest_start_addr, GUEST_SIZE)]);

/// End address - 0x10
let buf_guest_addr = GuestAddress(GUEST_START + GUEST_SIZE as u64 - 0x10);

/// Slice is to large
let buf = &mut [0u8; 0x100];

let result = guest_memory_mmap.find_region(addr).unwrap().write_slice(buf, buf_guest_addr);
```

### Why does the first one fails while the second succeeds?

Not sure how this would build. How do you call `write_slice` on a `GuestMemoryMmap` ?:
```Rust
let result = guest_memory_mmap.write_slice(buf, addr);
```

Also, did you mean `write` and not `write_slice`?

Reply:
nm, `GuestMemoryMmap` is a `GuestMemory`, who implements `Bytes`. So calling `write_slice` on it would make sense. But I still wonder if you meant `write` instead of `write_slice`...

Reply:
It's a special behavior to be backward compatible with firecracker, which fails when a write_slice() request crosses the memory region boundary. So the first example shows the way maintain current firecracker behavior. If it causes confusion, I will remove it.

Reply:
Will improve the examples for readability:)